---
title: "MOFA+: analysis of matching scRNA-seq and scATAC-seq data"
author:
  name: "Ricard Argelaguet"
  affiliation: "European Bioinformatics Institute, Cambridge, UK"
  email: "ricard@ebi.ac.uk"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{MOFA2: Applying MOFA2 to SNARE-seq data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Description

This vignette demonstrates how MOFA can be used to integrate scRNA-seq and scATAC-seq data from the same cell.  
There are very few protocols to date that enable this, one of them is [SNARE-seq](https://www.nature.com/articles/s41587-019-0290-0). As a demonstration we will use a simple data set of ~1000 cells where they mixed four cell lines together: BJ, GM12878, H1 and K562.  
In this setting, MOFA+ should be able to detect the (coordinated) variability in the RNA expression and in chromatin accessibiliy that determines the four different cell types.

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, fig.align="center")
```

# Load libraries

Make sure that `MOFA2` is imported last, to avoid collisions with functions from other packages
```{r, message=FALSE}
library(data.table)
library(purrr)
library(Seurat)
library(ggplot2)
library(gridExtra)
library(MOFA2)
```

## (Optional) Set up reticulate connection
```{r}
# reticulate::use_python("/Users/ricard/anaconda3/envs/base_new/bin/python", required = T)
```

# Load data

Load RNA modality as Seurat object
```{r}
seurat_rna <- readRDS(url("ftp://ftp.ebi.ac.uk/pub/databases/mofa/snare_seq/seurat_rna.rds"))
dim(seurat_rna)
```

Load ATAC modality as Seurat object
```{r }
seurat_atac <- readRDS(url("ftp://ftp.ebi.ac.uk/pub/databases/mofa/snare_seq/seurat_atac.rds"))
dim(seurat_atac)
```

Load metadata
```{r}
metadata <- fread("ftp://ftp.ebi.ac.uk/pub/databases/mofa/snare_seq/cell_metadata.txt", header=T)
```

Let's make sure that we subset cells that match in both assays:
```{r }
barcodes <- intersect(colnames(seurat_atac), colnames(seurat_rna))
seurat_atac <- seurat_atac[,barcodes]
seurat_rna <- seurat_rna[,barcodes]
```

## Feature selection

Subset top 2500 most variable genes and top 10000 most variable ATAC peaks
```{r message=FALSE }
# RNA
seurat_rna <- FindVariableFeatures(seurat_rna, 
  selection.method = "vst", 
  nfeatures = 2500
)
rna.features <- seurat_rna@assays$RNA@var.features

# ATAC
seurat_atac <- FindVariableFeatures(seurat_atac, 
  selection.method = "disp", 
  nfeatures = 5000
)
atac.features <- seurat_atac@assays$peaks@var.features
```

# Train MOFA with ATAC-seq peaks

## Create MOFA object

```{r}
mofa <- create_mofa(list(
  "RNA" = as.matrix( seurat_rna@assays$RNA@data[rna.features,] ),
  "ATAC" = as.matrix( seurat_atac@assays$peaks@data[atac.features,] )
))

mofa
```

## Define MOFA options

```{r}
# Model options: let's use only 4 factors, should be enough to distinguish the four cell lines.
model_opts <- get_default_model_options(mofa)
model_opts$num_factors <- 4

# Training options: let's use default options
train_opts <- get_default_training_options(mofa)
train_opts$seed <- 42
```

## Prepare the MOFA object

```{r}
mofa <- prepare_mofa(
  object = mofa,
  model_options = model_opts,
  training_options = train_opts
)
```

## Run MOFA

```{r message=FALSE}
mofa <- run_mofa(mofa)
```


## Downstream analysis

### Add cell metadata to the model

```{r, echo=FALSE}
metadata_to_mofa <- metadata %>%
  setnames("cell","sample") %>%   # add "sample" column
  setkey(sample) %>% .[unlist(samples_names(mofa))]   # sort data frame to match order of samples

samples_metadata(mofa) <- metadata_to_mofa
```

### Variance decomposition

Plot variance explained by factor
```{r}
plot_variance_explained(mofa)
```

Plot total variance explained per view
```{r}
plot_variance_explained(mofa, plot_total = TRUE)[[2]]
```

In general, we observe that most of the variation the model captures is driven by the RNA expression and little variance is explained by the ATAC data. Why? We will explore below. First let's have a look  whether the factors make sense.

### Plot factors

Looks like indeed the MOFA factors are capturing cell line variation
```{r message=FALSE}
plot_factors(mofa, factors=1:4, color_by = "cell_line")
```

Let's confirm this by plotting a UMAP based on the MOFA factors:
```{r cache=TRUE}
mofa <- run_umap(mofa)
plot_dimred(mofa, method="UMAP", color_by = "cell_line")
```

### Plot weights

Plot distribution of RNA weights per factor, highlighting the top ones
```{r message=FALSE}
plot_weights(mofa, 
  view = "RNA", 
  factors = 1:2, 
  nfeatures = 6, 
  text_size = 4
)
```


### Plot data

Before denoising (plotting the actual input data)

```{r }
plot_data_heatmap(mofa, 
  view = "RNA", 
  factor = 1, 
  features = 20,
  show_rownames = T, show_colnames = F, 
  cluster_rows = T, cluster_cols = F,
  annotation_samples = "cell_line"
)
```

After denoising (plotting model predictions)
```{r }
plot_data_heatmap(mofa, 
  view = "RNA", 
  factor = 1, 
  features = 20,
  show_rownames = T, show_colnames = F, 
  cluster_rows = T, cluster_cols = F,
  annotation_samples = "cell_line",
  denoise = TRUE
)
```


### What is going on with the ATAC data?

RNA looks great, but what is going on with the ATAC data?

Let's plot a histogram of the data. 
```{r}
hist(mofa@data$ATAC[[1]])
```

```{r}
mean(mofa@data$ATAC[[1]]>0)
```
Wow, this is massively sparse. Only 0.5% of values are different than zero. No wonder why the model is uncapable of explaining more variance.  

Yet, based on the variance explained estimates it seems that a bit of signal is indeed recovered. Let's explore it. First, we plot the variance explained values for the top 25 ATAC peaks from every factor.
```{r}
# Fetch top 25 features per factor, sorted by absolute value of the weight
weights.dt <- get_weights(mofa, views="ATAC", factors="all", as.data.frame = T) %>%
  as.data.table %>%
  .[,abs_value:=abs(value)] %>%
  .[,.SD %>% setorder(abs_value) %>% tail(n=25), by="factor"]
top_features <- weights.dt[,feature] %>% as.character %>% unique

# Fetch variance explained values per feature
variance.dt <- plot_variance_explained_per_feature(mofa, 
  view = "ATAC",
  features = top_features, 
  return_data = T
) %>% as.data.table

# Merge
to.plot <- merge(
  variance.dt[,c("feature","value")], 
  weights.dt[,c("feature","factor")], by="feature")
```

For the features with the largest weight (per factor) the model is capturing a bit of signal, but it is indeed very small. At most only 5% of the variance is explained:
```{r}
ggplot(to.plot, aes(x=feature, y=value*100)) +
  geom_bar(stat="identity") +
  facet_wrap(~factor, scales="free_x") +
  labs(x="ATAC peaks", y="Variance explained (%)") +
  theme_classic() +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )
```

As we did before with the RNA, let's visualise the actual data (figure below). 
```{r}
plot_data_heatmap(mofa, 
  view = "ATAC", 
  factor = 1, 
  features = 20,
  show_rownames = T, show_colnames = F, 
  cluster_rows = T, cluster_cols = F,
  annotation_samples = "cell_line"
)
```

If we plot the (denoised) model predictions this actually looks a bit better, but the input data is just of very poor quality.
```{r}
plot_data_heatmap(mofa, 
  view = "ATAC", 
  factor = 1, 
  features = 20,
  show_rownames = T, show_colnames = F, 
  cluster_rows = T, cluster_cols = F,
  annotation_samples = "cell_line",
  denoise = TRUE
)
```


# Train MOFA replacing the sparse ATAC matrix by a cisTopic embeddings

[cisTopic](https://github.com/aertslab/cisTopic) is a method for the simultaneous identification of cis-regulatory topics and cell states from single cell ATAC-seq data. The underlying Latent Dirichlet Allocation (LDA) model is a latent variable model (i.e. analogous to factor analysis) that is very well suited to deal with the sparsity of scATAC-seq data. Similar to MOFA, cisTopic extracts cell and feature (i.e. peaks) embeddings that can be used for downstream analysis.  

Here we will use cisTopic to create a (cell,topics) matrix that we will use as input to MOFA.

```{r}
library(cisTopic)
```

## Create cisTopic object

```{r}
cisTopicObject <- createcisTopicObject(
  count.matrix = seurat_atac@assays$peaks@counts, 
  project.name = 'CellLineMixture'
)
cisTopicObject
```

## Train cisTopic models

```{r, eval=FALSE}
topics <- c(10, 15, 20, 25)

cisTopicObject <- runCGSModels(
  cisTopicObject,
  topic = topics,
  burnin = 250,
  iterations = 500,
  nCores = 2
)

# Model selection
cisTopicObject <- selectModel(cisTopicObject)
```

Load pre-computed solution
```{r}
cisTopicObject <- readRDS("/Users/ricard/data/Chen2019/CellLineMixture/atac/cistopics.rds")
```

## Extract topic matrix

```{r}
cistopics_embeddings <- modelMatSelection(
  cisTopicObject, 
  target = "cell", 
  method = "Z-score"
)

# Sort samples to match same order as in the RNA
cistopics_embeddings <- cistopics_embeddings[,colnames(seurat_rna)]

dim(cistopics_embeddings)
```

## Create and run MOFA object 

Notice that we replace the ATAC peak matrix by the cisTopic cell embedding.
```{r}
mofa <- create_mofa(list(
  "RNA" = as.matrix( seurat_rna@assays$RNA@data[rna.features,] ),
  "ATAC" = cistopics_embeddings
))

mofa
```

Define MOFA options
```{r}
# Model options
model_opts <- get_default_model_options(mofa)
model_opts$num_factors <- 4

# Prepare MOFA object
mofa <- prepare_mofa(
  object = mofa,
  model_options = model_opts
)
```

Train the model
```{r cache=TRUE}
mofa <- run_mofa(mofa)
```

## Downstream analysis

### Add cell metadata to MOFA
```{r, echo=FALSE}
samples_metadata(mofa) <- metadata_to_mofa
```

### Variance decomposition

Plot variance explained per factor
```{r}
plot_variance_explained(mofa)
```

Plot total variance explained per view
```{r}
plot_variance_explained(mofa, plot_total = TRUE)[[2]]
```

This looks much better now! a lot of variance is explained by the ATAC topics. The reason is that the topics provide a compressed and denoised representation of the ATAC-seq data based on the Latent Dirichlet Allocation model. 

### Plot factors

Let's again plot a UMAP based on the MOFA factors to confirm that the factors capture the cell line variability:

```{r }
mofa <- run_umap(mofa)

plot_dimred(mofa, method="UMAP", color_by = "cell_line")
```

Plot UMAP coloured by the z-score values (per cell) for some topics. Some nice signal in there!
```{r}
p1 <- plot_dimred(mofa, method="UMAP", color_by = "Topic1")
p2 <- plot_dimred(mofa, method="UMAP", color_by = "Topic3")
p3 <- plot_dimred(mofa, method="UMAP", color_by = "Topic7")
p4 <- plot_dimred(mofa, method="UMAP", color_by = "Topic13")
grid.arrange(p1, p2, p3, p4, nrow= 2, ncol = 2)
```

### Plot weights

```{r message=FALSE}
plot_weights(mofa, 
  view = "ATAC", 
  factors = 1:3, 
  nfeatures = 3, 
  text_size = 5, 
  dot_size = 1.5
)
```

### Plot data

```{r }
plot_data_heatmap(mofa, 
  view = "ATAC", 
  factor = 1, 
  features = 100,
  show_rownames = T, show_colnames = F, 
  cluster_rows = T, cluster_cols = F,
)
```



## How to characterise the topics?

There are multiple ways of characterising the topics. We'll show a couple of examples, but we recommend you to visit the [cisTopic tutorials](https://github.com/aertslab/cisTopic).

### Inspect cisTopic output

cisTopic returns two distributions that represent: (1) the topic contributions per cell and (2) the region contribution to a topic. As in MOFA, you can fetch the corresponding values and try interpretring the etiology of each topic.  

To analyze the regions, the first step is always to derive a score that evaluates how likely is for a region to belong to a topic. This is done using `getRegionsScores()`. These scores can be rescaled into the range [0,1], which will be useful for the susbsequent binarization step:
```{r}
cisTopicObject <- getRegionsScores(cisTopicObject, method = "NormTop", scaled = TRUE)
cisTopicObject <- binarizecisTopics(cisTopicObject, plot = FALSE)
```

Extract region loadings (in z-scores)
```{r}
loadings <- modelMatSelection(cisTopicObject, target = "region", method = "Z-score")
hist(loadings)
```

```{r}
loadings[1:5,1:3]
```


### Annotation to genes and GO terms

```{r, message=FALSE}
library(ChIPseeker)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)

# Add peak annotation to the regions (e.g. type of region, closest gene)
cisTopicObject <- annotateRegions(
  cisTopicObject, 
  txdb = TxDb.Hsapiens.UCSC.hg38.knownGene, 
  annoDb = 'org.Hs.eg.db'
)
```

Heatmap containing the row normalised AUC values for the signatures in the topics
```{r, message=FALSE}
signaturesHeatmap(
  cisTopicObject, 
  selected.signatures = 'annotation'
)
```


### Transcription Factor motif enrichment

```{r}
# cisTopicObject <- binarizedcisTopicsToCtx(cisTopicObject, genome='hg19')
# 
# cisTopicObject <- scoredRegionsToCtx(cisTopicObject, genome='hg19')
# 
# pathToFeather <- "hg19-regions-9species.all_regions.mc8nr.feather"
# 
# cisTopicObject <- topicsRcisTarget(cisTopicObject, genome='hg19', pathToFeather, reduced_database=FALSE, nesThreshold=3, rocthr=0.005, maxRank=20000, nCores=5)
# 
# Topic6_motif_enr <- cisTopicObject@binarized.RcisTarget[[6]]
# 
# DT::datatable(Topic6_motif_enr[,-c("enrichedRegions", "TF_lowConf"), with=FALSE], escape = FALSE, filter="top", options=list(pageLength=5))
```


# Conclusion

The count ATAC-seq matrix is too sparse to be used as input to MOFA. We recommend applying some pre-processing options to enrich the signal. A simple solution could be to aggregate peaks and summarise them. The second option, which we demonstrated here, is to use a signal extraction method such as cis-Topics, and then take their output as input to MOFA.

# Acknowledgments 
I want to thank Carmen Bravo for feedback on how to best use CisTopic, and Song Chen for sharing the scripts to analyse SNARE-seq data.  
Having said this, one should not request scripts to reproduce results by e-mail in 2020, we should all document our code and release it upon publication.

<details>
  <summary>**Session Info**</summary>
  
```{r}
sessionInfo()
```

</details>
